pkg/pki/pki.go

```diff
@@ -0,0 +1,189 @@
+/* -------------------------------------------------------------------------- *\
+ *             Apache 2.0 License Copyright © 2022 The Aurae Authors          *
+ *                                                                            *
+ *                +--------------------------------------------+              *
+ *                |   █████╗ ██╗   ██╗██████╗  █████╗ ███████╗ |              *
+ *                |  ██╔══██╗██║   ██║██╔══██╗██╔══██╗██╔════╝ |              *
+ *                |  ███████║██║   ██║██████╔╝███████║█████╗   |              *
+ *                |  ██╔══██║██║   ██║██╔══██╗██╔══██║██╔══╝   |              *
+ *                |  ██║  ██║╚██████╔╝██║  ██║██║  ██║███████╗ |              *
+ *                |  ╚═╝  ╚═╝ ╚═════╝ ╚═╝  ╚═╝╚═╝  ╚═╝╚══════╝ |              *
+ *                +--------------------------------------------+              *
+ *                                                                            *
+ *                         Distributed Systems Runtime                        *
+ *                                                                            *
+ * -------------------------------------------------------------------------- *
+ *                                                                            *
+ *   Licensed under the Apache License, Version 2.0 (the "License");          *
+ *   you may not use this file except in compliance with the License.         *
+ *   You may obtain a copy of the License at                                  *
+ *                                                                            *
+ *       http://www.apache.org/licenses/LICENSE-2.0                           *
+ *                                                                            *
+ *   Unless required by applicable law or agreed to in writing, software      *
+ *   distributed under the License is distributed on an "AS IS" BASIS,        *
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ *   See the License for the specific language governing permissions and      *
+ *   limitations under the License.                                           *
+ *                                                                            *
+\* -------------------------------------------------------------------------- */
+
+package pki
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+type AuraeCA struct {
+	Certificate string `json:"cert" yaml:"cert"`
+	Key         string `json:"key" yaml:"key"`
+}
+
+func CreateAuraeRootCA(path string, domainName string) (*AuraeCA, error) {
```

Comment:
Is this an Aurae specific root CA? or is this just a normal root CA that we happen to be using for Aurae?

Basically the function name `CreateEntity` makes me think that our entity is a **AuraeRootCA** which makes me think we have special things here that are unique to the project. Is this true? Or is this just a normal CA?

Reply:
I thought about this as well and came to the following conclusion:

We are currently hardcoding all the subject names with something `aurae*` terms. The resulting certificate always looks like this:

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            6a:d9:8a:7e:a8:81:24:5d:07:af:43:6e:fa:99:03:56
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: C = IS, L = aurae, street = aurae, O = Aurae, OU = Runtime, CN = my.domain.com
        Validity
            Not Before: Jan 13 07:58:01 2023 GMT
            Not After : May 30 07:58:01 2050 GMT
        Subject: C = IS, L = aurae, street = aurae, O = Aurae, OU = Runtime, CN = my.domain.com
```

As long as we have no option changing those names (e.g. by providing a csr config as mentioned above) I think it is reasonable to stick to the naming of "createAuraeCA". In the ned we have a root ca that will always contain the "aurae-flavor" :D

I'd therefore advocate to change the naming *if* we ever plan to change this behavior.